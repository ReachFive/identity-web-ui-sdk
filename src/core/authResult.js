import { parseJwtTokenPayload } from '../helpers/jwt'
import { logError } from '../helpers/logger'

export function enrichAuthResult(response) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value) {
    return value && (value.accessToken || value.idToken);
}

export function isErrorResponse(value) {
    return value && value.error;
}
