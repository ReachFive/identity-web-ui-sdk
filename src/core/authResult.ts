import { AuthResult, ErrorResponse } from '@reachfive/identity-core';

import { parseJwtTokenPayload } from '../helpers/jwt';
import { logError } from '../helpers/logger';

export function enrichAuthResult(response: AuthResult) {
    if (response.idToken) {
        try {
            const idTokenPayload = parseJwtTokenPayload(response.idToken);

            return {
                ...response,
                idTokenPayload,
            };
        } catch (e) {
            logError('id token parsing error: ' + e);
        }
    }

    return response;
}

export function isAuthResult(value: unknown): value is AuthResult {
    return (
        value !== null &&
        typeof value === 'object' &&
        ('accessToken' in value || 'idToken' in value)
    );
}

export function isErrorResponse(value: unknown): value is ErrorResponse {
    return value !== null && typeof value === 'object' && 'error' in value;
}
