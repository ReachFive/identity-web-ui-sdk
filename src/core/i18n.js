import isPlainObject from 'lodash-es/isPlainObject';

export default function resolveI18n(defaultMessages, messages = {}) {
    const mergedMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key, params) => {
        const template = mergedMessages[key] || key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param]), template)
            : template;
    };
}

function flattenObject(object, prefix = []) {
    return isPlainObject(object)
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
