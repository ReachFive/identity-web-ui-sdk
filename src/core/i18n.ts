export type I18nMessages = Record<string, string>

export type I18nNestedMessages = Record<string, string | I18nMessages>

export type I18nMessageParams = Record<string, unknown>

export type I18nResolver = (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => string

export function resolveI18n<T extends (Record<string, T> | string)>(defaultMessages: I18nMessages = {}, messages: I18nNestedMessages = {}): I18nResolver {
    const mergedMessages: I18nMessages = {
        ...defaultMessages,
        ...flattenObject(messages)
    };

    return (key: string, params?: I18nMessageParams, fallback?: (params?: I18nMessageParams) => string) => {
        const template = mergedMessages[key] ?? fallback?.(params) ?? key;

        return params
            ? Object.keys(params).reduce((acc, param) => acc.replace(`{${param}}`, params[param] as string), template)
            : template;
    };
}

export default resolveI18n

function flattenObject(object: I18nNestedMessages | string, prefix: string[] = []): Record<string, string> {
    return typeof object === "object" && object !== null
        ? Object.keys(object).reduce((acc, key) => ({
            ...acc,
            ...flattenObject(object[key], [...prefix, key])
        }), {})
        : ({ [prefix.join('.')]: object });
}
