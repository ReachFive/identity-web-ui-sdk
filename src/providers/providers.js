export { default as facebook } from './facebook/facebook'
