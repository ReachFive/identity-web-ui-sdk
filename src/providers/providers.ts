export interface windowSize {
    width: number;
    height: number;
}

export interface Provider {
    key: string;
    name: string;
    color: string;
    fontWeight?: number;
    fontFamily?: string;
    fontKey?: string;
    btnBorderColor?: string;
    btnTextColor?: string;
    btnBackgroundColor?: string;
    icon: string;
    windowSize?: windowSize;
    hidden?: boolean;
}

import { default as amazon } from './amazon/amazon';
import { default as apple } from './apple/apple';
import { default as bconnect } from './bconnect/bconnect';
import { default as facebook } from './facebook/facebook';
import { default as franceconnect } from './franceconnect/franceconnect';
import { default as google } from './google/google';
import { default as kakaotalk } from './kakaotalk/kakaotalk';
import { default as line } from './line/line';
import { default as linkedin } from './linkedin/linkedin';
import { default as mailru } from './mailru/mailru';
import { default as microsoft } from './microsoft/microsoft';
import { default as naver } from './naver/naver';
import { default as okta } from './okta/okta';
import { default as oney } from './oney/oney';
import { default as orange } from './orange/orange';
import { default as paypal } from './paypal/paypal';
import { default as ping } from './ping/ping';
import { default as qq } from './qq/qq';
import { default as tiktok } from './tiktok/tiktok';
import { default as twitter } from './twitter/twitter';
import { default as vkontakte } from './vkontakte/vkontakte';
import { default as wechat } from './wechat/wechat';
import { default as weibo } from './weibo/weibo';
import { default as yandex } from './yandex/yandex';

export const providers = {
    amazon: amazon,
    apple: apple,
    bconnect: bconnect,
    facebook: facebook,
    franceconnect: franceconnect,
    google: google,
    kakaotalk: kakaotalk,
    line: line,
    linkedin: linkedin,
    mailru: mailru,
    microsoft: microsoft,
    naver: naver,
    okta: okta,
    oney: oney,
    orange: orange,
    paypal: paypal,
    ping: ping,
    qq: qq,
    tiktok: tiktok,
    twitter: twitter,
    vkontakte: vkontakte,
    wechat: wechat,
    weibo: weibo,
    yandex: yandex,
};

export default providers;

export type ProviderId = keyof typeof providers;
