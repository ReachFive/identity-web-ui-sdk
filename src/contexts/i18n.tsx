import React, { ComponentType, PropsWithChildren, useMemo } from 'react';

import { I18nMessages, I18nNestedMessages, I18nResolver, resolveI18n } from '../core/i18n';

export interface Props {
  defaultMessages?: I18nMessages
  messages?: I18nNestedMessages
}

export const I18nContext = React.createContext<I18nResolver | undefined>(undefined);

export function useI18n(): I18nResolver {
  const context = React.useContext(I18nContext);
  if (!context) {
    throw new Error('No I18nContext provided');
  }

  return context;
}

export interface I18nProps {
  i18n: I18nResolver
}

export type WithI18n<P> = P & I18nProps

export function withI18n<T extends I18nProps = I18nProps>(WrappedComponent: ComponentType<T>) {
  const displayName = WrappedComponent.displayName || WrappedComponent.name || "Component";

  const ComponentWithI18n = (props: Omit<T, keyof I18nProps>) => {
    const i18n = useI18n();
    return <WrappedComponent {...({ i18n })} {...(props as T)} />;
  };

  ComponentWithI18n.displayName = `withI18n(${displayName})`;

  return ComponentWithI18n;
}

export function I18nProvider({ children, defaultMessages, messages }: PropsWithChildren<Props>): JSX.Element | null {
  const resolver = useMemo(() => resolveI18n(defaultMessages, messages), [defaultMessages, messages])
  return <I18nContext.Provider value={resolver}>{children}</I18nContext.Provider>
}
