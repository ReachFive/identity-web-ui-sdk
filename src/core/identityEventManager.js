import { enrichAuthResult } from './authResult'
import EventManager from '../helpers/eventManager'

export function createEventManager() {
    const eventManager = new EventManager();

    return {
        on(eventName, listener) {
            eventManager.on(eventName, listener);
        },

        off(eventName, listener) {
            eventManager.off(eventName, listener)
        },

        fireEvent(eventName, data) {
            if (eventName === 'authenticated') {
                eventManager.fire(eventName, enrichAuthResult(data));
            } else {
                eventManager.fire(eventName, data);
            }
        }
    }
}
