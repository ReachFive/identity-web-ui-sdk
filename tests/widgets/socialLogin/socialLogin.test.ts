/**
 * @jest-environment jest-fixed-jsdom
 */

import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import '@testing-library/jest-dom/jest-globals';
import { screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import 'jest-styled-components';

import { type Client } from '@reachfive/identity-core';

import { type I18nMessages } from '../../../src/core/i18n';
import { providers, type ProviderId } from '../../../src/providers/providers';

import { OnError, OnSuccess } from '@/types';
import SocialLoginWidget from '../../../src/widgets/socialLogin/socialLoginWidget';
import { componentGenerator, defaultConfig, snapshotGenerator } from '../renderer';

const defaultI18n: I18nMessages = {};

describe('Snapshot', () => {
    // @ts-expect-error partial Client
    const apiClient: Client = {};

    const generateSnapshot = snapshotGenerator(SocialLoginWidget, apiClient, defaultI18n);

    describe('social login', () => {
        test('basic', generateSnapshot({}));
    });
});

describe('DOM testing', () => {
    const loginWithSocialProvider = jest.fn<Client['loginWithSocialProvider']>();

    const onError = jest.fn<OnError>();
    const onSuccess = jest.fn<OnSuccess>();

    beforeEach(() => {
        loginWithSocialProvider.mockClear();
        onError.mockClear();
        onSuccess.mockClear();
    });

    // @ts-expect-error partial Client
    const apiClient: Client = {
        loginWithSocialProvider,
    };

    const generateComponent = componentGenerator(SocialLoginWidget, apiClient, defaultI18n);

    test('basic', async () => {
        const user = userEvent.setup();

        loginWithSocialProvider.mockResolvedValue(undefined);

        await generateComponent({ onError, onSuccess });

        defaultConfig.socialProviders.forEach(provider => {
            expect(screen.queryByTitle(providers[provider as ProviderId].name)).toBeInTheDocument();
        });

        const provider = defaultConfig.socialProviders[0] as ProviderId;
        const button = screen.getByTitle(providers[provider].name);
        await user.click(button);

        expect(loginWithSocialProvider).toBeCalledWith(provider, undefined);

        expect(onSuccess).toBeCalledWith(
            expect.objectContaining({
                name: 'login',
                authResult: expect.objectContaining({ providerName: 'facebook' }),
            })
        );
        expect(onError).not.toBeCalled();
    });

    test('themed', async () => {
        await generateComponent({
            theme: {
                primaryColor: '#ff0000',
            },
            onError,
            onSuccess,
        });

        defaultConfig.socialProviders.forEach(provider => {
            expect(screen.queryByTitle(providers[provider as ProviderId].name)).toBeInTheDocument();
        });
    });

    test('login with social failure', async () => {
        const user = userEvent.setup();

        const error = { error: 'Unexpected error' };
        loginWithSocialProvider.mockRejectedValue(error);

        await generateComponent({ onError, onSuccess });

        defaultConfig.socialProviders.forEach(provider => {
            expect(screen.queryByTitle(providers[provider as ProviderId].name)).toBeInTheDocument();
        });

        const provider = defaultConfig.socialProviders[0] as ProviderId;
        const button = screen.getByTitle(providers[provider].name);
        await user.click(button);

        await waitFor(() => expect(loginWithSocialProvider).toBeCalledWith(provider, undefined));

        expect(onSuccess).not.toBeCalled();
        expect(onError).toBeCalledWith(error);
    });
});
