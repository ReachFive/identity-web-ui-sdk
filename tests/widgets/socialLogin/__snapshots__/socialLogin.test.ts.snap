// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot social login basic 1`] = `
.c0 {
  font-size: 14px;
  -webkit-transition: -webkit-transform 400ms ease, opacity 400ms ease;
  -webkit-transition: transform 400ms ease, opacity 400ms ease;
  transition: transform 400ms ease, opacity 400ms ease;
  opacity: 0;
}

.c3 {
  position: absolute;
  left: 0;
  top: 0;
  bottom: 0;
  width: 38px;
  box-sizing: border-box;
  border-radius: 2px;
  background-image: url(svg);
  background-repeat: no-repeat;
  background-size: 20px 20px;
  background-position: center center;
}

.c2 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #ffffff;
  background-color: #3b5998;
  border: 1px solid #3b5998;
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
  margin-bottom: 10px;
  position: relative;
  width: 100%;
  height: 40px;
}

.c2:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(59,89,152,0.5);
}

.c2:hover,
.c2:active {
  color: #ffffff;
  background-color: #30487b;
  border-color: #30487b;
}

.c2[disabled] {
  opacity: .65;
}

.c4 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  text-align: center;
  font-weight: bold;
  vertical-align: middle;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  touch-action: manipulation;
  cursor: pointer;
  color: #58666e;
  background-color: #ffffff;
  border: 1px solid rgba(0,0,0,.15);
  padding: 9px 12px;
  font-size: 14px;
  line-height: 1.428571429;
  border-radius: 3px;
  -webkit-transition: all .15s ease-in-out;
  transition: all .15s ease-in-out;
  margin-bottom: 10px;
  position: relative;
  width: 100%;
  height: 40px;
}

.c4:focus {
  outline: 0;
  box-shadow: 0 0 0 3px rgba(0,0,0,0);
}

.c4:hover,
.c4:active {
  color: #58666e;
  background-color: #ebebeb;
  border-color: rgba(0,0,0,0.15);
}

.c4[disabled] {
  opacity: .65;
}

.c1 {
  text-align: center;
}

<div>
  <div
    class="c0 r5-social-login r5-widget-active"
  >
    <div
      class="r5-social-buttons c1"
    >
      <button
        class="c2 r5-btn-social r5-btn-social-facebook"
        title="Facebook"
      >
        <span
          class="r5-btn-social-icon c3"
        />
        <span
          class="r5-btn-social-text"
        >
          Facebook
        </span>
      </button>
      <button
        class="c4 r5-btn-social r5-btn-social-google"
        title="Google"
      >
        <span
          class="r5-btn-social-icon c3"
        />
        <span
          class="r5-btn-social-text"
        >
          Google
        </span>
      </button>
    </div>
  </div>
</div>
`;
